lib/MongoDB/Role/_SessionSupport.pm

```diff
@@ -92,6 +92,20 @@ sub _update_operation_time {
     return;
 }
 
+# Certain errors have to happen as soon as possible, such as write concern
+# errors in a retryable write. This has to be seperate to the other functions
+# due to not all result objects having the base response inside, so cannot be
+# used to parse operationTime or $clusterTime
+sub _assert_session_errors {
```

Comment:
I'm a little nervous about this, but if all existing tests pass (particularly for bulk operations) then I'm okay with it.

Reply:
Yea this was the least complicated place to add this assertion - I couldn't see any information about the comment in `MongoDB::Op::_BulkWrite->execute` of `write concern errors are thrown only for the entire batch` in the specs for bulk operations, and as retryable writes can happen on certain write concern errors it must be done inside the retry loop.

The other single ops do throw WriteConcernErrors further down the stack, but only on the return value of the `send_*write_op` which is outside the retryable writes block - could maybe move that assertion up the call stack slightly to be inside the execute block? If commandResult is also modified to check for write concern errors in the assert step, that would get rid of this extra assertion? Would mean that bulk writes would throw write concern errors earlier, and as I said I couldnt find if that is a spec decision or not...